### DISCUSSION TITLE
Open wrt support

### DISCUSSION THREAD

**Discussion by ANARHIST1984** [Koenkk/zigbee2mqtt#6717] [General]
I try to install Zigbee2Mqtt to Openwrt router. 
openwrt_19_07_7_ramips_mt7621_zbtlink_zbt_we3526
I manage to compile from here https://github.com/nxhack/openwrt-node-packages/tree/dev_19.07
node_v12.20.0, node-npm_v12.20.0, node-zigbee2mqtt_1.17.1.
And just in case node-bcrypt_5.0.1, node-serialport_9.0.6, node-serialport-bindings_9.0.7, node-serialport-list_9.0.4.
also install driver kmod-usb-acm and i can see in system /dev/ttyACM0

when i try to start "npm start" i receive this
root@OpenWrt:/opt/zigbee2mqtt# npm start
npm info it worked if it ends with ok
npm info using npm@6.14.8
npm info using node@v14.15.1
npm info lifecycle zigbee2mqtt@1.16.2~prestart: zigbee2mqtt@1.16.2
npm info lifecycle zigbee2mqtt@1.16.2~start: zigbee2mqtt@1.16.2

> zigbee2mqtt@1.16.2 start /opt/zigbee2mqtt
> node index.js

Zigbee2MQTT:info  2021-03-14 22:43:56: Logging to console and directory: '/opt/zigbee2mqtt/data/log/2021-03-14.22-43-49' filename: log.txt
Zigbee2MQTT:info  2021-03-14 22:43:57: Starting Zigbee2MQTT version 1.16.2 (commit #unknown)
Zigbee2MQTT:info  2021-03-14 22:43:57: Starting zigbee-herdsman (0.13.37)
Zigbee2MQTT:error 2021-03-14 22:43:57: Error while starting zigbee-herdsman
Zigbee2MQTT:error 2021-03-14 22:43:57: Failed to start zigbee
Zigbee2MQTT:error 2021-03-14 22:43:57: Exiting...
Zigbee2MQTT:error 2021-03-14 22:43:57: Error: Error relocating /opt/zigbee2mqtt/node_modules/zigbee-herdsman/node_modules/@serialport/bindings/build/Release/bindings.node: _ZN2v82V88MakeWeakEPPNS_8internal6ObjectEPvPFvRKNS_16WeakCallbackInfoIvEEENS_16WeakCallbackTypeE: symbol not found
    at Object.Module._extensions..node (internal/modules/cjs/loader.js:1122:18)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at bindings (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/node_modules/bindings/bindings.js:112:48)
    at Object.<anonymous> (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/node_modules/@serialport/bindings/lib/linux.js:2:36)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
npm info lifecycle zigbee2mqtt@1.16.2~start: Failed to exec start script
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! zigbee2mqtt@1.16.2 start: `node index.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the zigbee2mqtt@1.16.2 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm timing npm Completed in 20808ms

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2021-03-14T22_43_58_259Z-debug.log

Log
 info it worked if it ends with ok
1 verbose cli [ '/usr/bin/node', '/usr/bin/npm', 'start' ]
2 info using npm@6.14.8
3 info using node@v14.15.1
4 verbose run-script [ 'prestart', 'start', 'poststart' ]
5 info lifecycle zigbee2mqtt@1.16.2~prestart: zigbee2mqtt@1.16.2
6 info lifecycle zigbee2mqtt@1.16.2~start: zigbee2mqtt@1.16.2
7 verbose lifecycle zigbee2mqtt@1.16.2~start: unsafe-perm in lifecycle true
8 verbose lifecycle zigbee2mqtt@1.16.2~start: PATH: /usr/lib/node_modules/npm/node_modu>
9 verbose lifecycle zigbee2mqtt@1.16.2~start: CWD: /opt/zigbee2mqtt
10 silly lifecycle zigbee2mqtt@1.16.2~start: Args: [ '-c', 'node index.js' ]
11 silly lifecycle zigbee2mqtt@1.16.2~start: Returned: code: 1  signal: null
12 info lifecycle zigbee2mqtt@1.16.2~start: Failed to exec start script
13 verbose stack Error: zigbee2mqtt@1.16.2 start: `node index.js`
13 verbose stack Exit status 1
13 verbose stack     at EventEmitter.<anonymous> (/usr/lib/node_modules/npm/node_module>
13 verbose stack     at EventEmitter.emit (events.js:315:20)
13 verbose stack     at ChildProcess.<anonymous> (/usr/lib/node_modules/npm/node_module>
13 verbose stack     at ChildProcess.emit (events.js:315:20)
13 verbose stack     at maybeClose (internal/child_process.js:1048:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:>
14 verbose pkgid zigbee2mqtt@1.16.2
15 verbose cwd /opt/zigbee2mqtt

what am I missing or doing wrong?

**Comment by danergo** [comment]
I believe your architecture is too slow to run this project.

Python and node are way too heavy for a router. I know it's not too much help, sorry for that. I'm fighting with a similar issue on a RPi0, which might be around your router in terms of performance.

One thing you might consider though. Although it seems feasible to add this Zigbee functionality to your router (it runs constantly, doesn't consume too much etc). However it's still your network device, and you might run into series issues: your neighbor buys a new Zigbee device and your Netflix will stall/or will stream at bad quality, as your router is busy with handling ZB stuff. 

Of course if you have a powerful plant as router, these situations might be far away, but a router hardware is always designed and optimized for behaving as a router (and not as a general purpose computer).

So your mileage may vary, I'd be interested in your results, as Pi0 suffers from a very similar error message during startup.

**Reply by ANARHIST1984** [reply]
Openwrt 19.07 - Z2M work perfect, 2 month work perfect
Openwrt 21.02 - node not working, developers remove mips flu emulator options
Openwrt use 25mb ram
Z2m use 50mb
I use mt7621 880x4 512mb ram 512mb nand
Even slow router with 128mb ram and 400mhz CPU enough to run z2m. 
for use z2m in 21.02 need ARM cpu

**Reply by danergo** [reply]
Sounds great, except one: Z2M consumes 2times more ram than the whole router OS, that's not too great IMHO :)

**Reply by ANARHIST1984** [reply]
that's why I stop use Z2M and now I have Domoticz+ZiGate
Zigate more stable and use less memory
